saleor/order/migrations/tasks/saleor3_13.py

```diff
@@ -1,36 +1,47 @@
 from ....celeryconf import app
 from ...models import OrderEvent
 
+from django.db import transaction
 from django.db.models import QuerySet
 
-# Batch size of size 1000 is about 1MB memory usage in task
-BATCH_SIZE = 1000
+# Batch size of size 5000 is about 1MB memory usage in task
+BATCH_SIZE = 5000
 
 
 def update_type_to_transaction_cancel_requested(qs: QuerySet[OrderEvent]):
-    qs.update(type="transaction_cancel_requested")
+    with transaction.atomic():
+        # lock the batch of objects
+        _events = list(qs.select_for_update(of=(["self"])))
+        qs.update(type="transaction_cancel_requested")
 
 
 @app.task
 def order_events_rename_transaction_void_events_task():
-    qs = OrderEvent.objects.filter(type="transaction_void_requested").order_by("-pk")
-    ids = qs.values_list("pk", flat=True)[:BATCH_SIZE]
-    qs = qs.filter(pk__in=ids)
+    events = OrderEvent.objects.filter(type="transaction_void_requested").order_by(
+        "-pk"
+    )
+    ids = events.values_list("pk", flat=True)[:BATCH_SIZE]
+    qs = OrderEvent.objects.filter(pk__in=ids)
 
     if ids:
         update_type_to_transaction_cancel_requested(qs)
         order_events_rename_transaction_void_events_task.delay()
 
 
 def update_type_to_transaction_charge_requested(qs: QuerySet[OrderEvent]):
-    qs.update(type="transaction_charge_requested")
+    with transaction.atomic():
+        # lock the batch of objects
+        _events = list(qs.select_for_update(of=(["self"])))
```

Comment:
cast to list and thus evaluation of extra qs can be avoided and moved to subquery which will be much faster, see here on how it's done:
https://github.com/saleor/saleor/blob/main/saleor/graphql/attribute/mutations/attribute_value_update.py#L83
It also does the batching and in your case I believe you can also skip`with transaction.atomic():` if you use that approach. Update operations are atomic and since we're gonna do things in subquery that will be the only statement.

Reply:
In the example you sent, we block all related products until all batches will be finished, I'm not sure if it's right. Let's consider the migration will take an hour, the access to those products will be blocked for this time AFAIK

Reply:
Also keep in mind that this task will called in a celery worker. There is a possibility that we will have a different tasks that will try to update the same objects, which means that we need to lock a batch of objects to be sure that DB will not raise the deadlock exception.
This is mandatory when you have more that one celery worker that processes the data